src/snowflake/cli/plugins/object/common.py

```diff
@@ -0,0 +1,78 @@
+from dataclasses import dataclass
+from typing import Optional
+
+import typer
+from click import ClickException
+from snowflake.cli.api.project.util import (
+    QUOTED_IDENTIFIER_REGEX,
+    UNQUOTED_IDENTIFIER_REGEX,
+    is_valid_identifier,
+    to_string_literal,
+)
+
+
+@dataclass
+class Tag:
+    name: str
+    value: str
+
+    def __post_init__(self):
+        if not is_valid_identifier(self.name):
+            raise ValueError("name of a tag must be a valid snowflake identifier")
+
+    def value_string_literal(self):
+        return to_string_literal(self.value)
+
+
+def _parse_tag(tag: str) -> Tag:
+    import re
+
+    identifier_pattern = re.compile(
+        f"(?P<tag_name>{UNQUOTED_IDENTIFIER_REGEX}|{QUOTED_IDENTIFIER_REGEX})"
+    )
+    value_pattern = re.compile(f"(?P<tag_value>.+)")
+    result = re.fullmatch(f"{identifier_pattern.pattern}={value_pattern.pattern}", tag)
+    if result is not None:
+        try:
+            return Tag(result.group("tag_name"), result.group("tag_value"))
+        except ValueError:
+            raise ClickException(
+                "tag must be in the format <name>=<value> where 'name' is a valid identifier and value is a string"
+            )
+    else:
+        raise ClickException(
+            "tag must be in the format <name>=<value> where 'name' is a valid identifier and value is a string"
+        )
```

Comment:
Custom exception could be used for this repeating exceptions.

```python
class TagError(ClickException):
    def __init__(self):
        super().__init__("tag must be in the format <name>=<value> where 'name' is a valid identifier and value is a string")
```


```suggestion
    if result is not None:
        try:
            return Tag(result.group("tag_name"), result.group("tag_value"))
        except ValueError:
            raise TagError()
    else:
        raise TagError()
```

Reply:
appreciate the notes!